examples/cryptocurrency-advanced/backend/src/migrations/tests.rs

```diff
@@ -0,0 +1,225 @@
+    let wallets: HashMap<_, _> = generate_random_wallets().take(WALLET_COUNT).collect();
+    let wallets_iter = wallets
+        .iter()
+        .map(|(name, balance)| (name.as_str(), *balance));
+
+    let old_version = OldService.artifact_id().version;
+    let mut test = MigrationTest::new(CryptocurrencyService, old_version);
+    test.setup(|fork| {
+        prepare_wallets(fork, wallets_iter.clone());
+    });
+
+    let schema = SchemaImpl::new(test.migrate().end_snapshot());
+    assert_state(&schema, wallets_iter);
+}
+
+fn init_testkit() -> TestKit {
+    TestKitBuilder::validator()
+        // Add old version of the service.
+        .with_default_rust_service(OldService)
+        // Add the artifact for the new version.
+        .with_migrating_rust_service(CryptocurrencyService)
+        .with_artifact(CryptocurrencyService.artifact_id())
+        // Add the supervisor service.
+        .with_rust_service(Supervisor)
+        .with_artifact(Supervisor.artifact_id())
+        .with_instance(Supervisor::simple())
+        .build()
+}
+
+#[test]
+fn test_with_full_lifecycle() {
+    const SERVICE_ID: InstanceId = OldService::INSTANCE_ID;
+    let mut testkit = init_testkit();
+
+    // Create some accounts using the old service.
+    let alice = KeyPair::random();
+    let bob = KeyPair::random();
+    let carol = KeyPair::random();
+    let txs = vec![
+        Old::create_wallet(&alice, SERVICE_ID, OldCreate::new("alice")),
+        Old::create_wallet(&bob, SERVICE_ID, OldCreate::new("bob")),
+        Old::create_wallet(&carol, SERVICE_ID, OldCreate::new("carol")),
+    ];
+    let block = testkit.create_block_with_transactions(txs);
+    assert!(block.errors.is_empty());
+
+    // Transfer value between two accounts.
+    let transfer = OldTransfer {
+        to: bob.public_key(),
+        amount: 16,
+        seed: 0,
+    };
+    let transfer = Old::transfer(&alice, SERVICE_ID, transfer);
+    let block = testkit.create_block_with_transaction(transfer);
+    assert!(block.errors.is_empty());
+
+    // Migrate the service. To do this, we first need to stop or freeze the service.
+    let proposal = ConfigPropose::immediate(0).freeze_service(SERVICE_ID);
+    let admin_keys = testkit.us().service_keypair();
+    let freeze_service = admin_keys.propose_config_change(SUPERVISOR_ID, proposal);
+    let block = testkit.create_block_with_transaction(freeze_service);
+    assert!(block.errors.is_empty());
+
+    // Then, we can launch the migration.
+    let migration_req = MigrationRequest {
+        new_artifact: CryptocurrencyService.artifact_id(),
+        service: OldService::INSTANCE_NAME.to_owned(),
+        deadline_height: Height(100),
+    };
+    let migration = admin_keys.request_migration(SUPERVISOR_ID, migration_req.clone());
+    let block = testkit.create_block_with_transaction(migration);
+    assert!(block.errors.is_empty());
+
+    // Since migration is performed in a background, it may take a little while.
```

Comment:
Interesting, can we use futures in this case?

Reply:
Could you provide pseudocode how futures can be used? All other testkit code is synchronous, so I'm not sure how well the use of futures will fit into it.

Reply:
Migration performs in the separate thread, isn't it? So in theory we can return the corresponding join handle or something like oneshot in the deploy_artifact method.